### PR TITLE
Replace regex validators by a faster equivalent implementation.

### PR DESCRIPTION
**Pull request by raiju**
## Before this PR

Compiling large amounts of conjure takes a lot of time; for example, here's a sample run of one internal tool:

```
./gradlew $project:compileIr --rerun-tasks  1.43s user 0.23s system 2% cpu 1:09.24 total
```
I.e. 69s total

Main contributors from that sample run:
```
:$project:rawIr | 21.871s |  
:$project:compileIr | 21.492s |  
```

## After this PR

==COMMIT_MSG==
Replace regex validators by a faster equivalent implementation.
==COMMIT_MSG==

```
./gradlew $project:compileIr --rerun-tasks  1.32s user 0.20s system 4% cpu 34.243 total
```
I.e. 34s total

Main contributors from that sample run:
```
:$project:rawIr | 9.597s |  
:$project:compileIr | 8.637s |
```

### CODE REVIEW

File: conjure-generator-common/src/main/java/com/palantir/conjure/CaseConverter.java [hunk 1]
```diff
@@ -20,14 +20,16 @@
...
+    public static final CachedPatternMatcher CAMEL_CASE_PATTERN =
+            CachedPatternMatcher.wrap(Pattern.compile("^[a-z]([A-Z]{1," + "2}[a-z0-9]|[a-z0-9])*[A-Z]?$"));
+    public static final CachedPatternMatcher KEBAB_CASE_PATTERN =
+            CachedPatternMatcher.wrap(Pattern.compile("^[a-z]((-[a-z]){1,2}[a-z0-9]|[a-z0-9])*(-[a-z])?$"));
+    public static final CachedPatternMatcher SNAKE_CASE_PATTERN =
+            CachedPatternMatcher.wrap(Pattern.compile("^[a-z]((_[a-z]){1,2}[a-z0-9]|[a-z0-9])*(_[a-z])?$"));
 
     private CaseConverter() {}
 
+    @SuppressWarnings("ImmutableEnumChecker")
```
Comment (raiju) [inline conjure-generator-common/src/main/java/com/palantir/conjure/CaseConverter.java:32]: Up for discussion; the `CachedPatternMatcher` is technically not immutable, so the enum validator is correctly shouting, but it functionally acts like an immutable interface.

File: conjure-generator-common/src/main/java/com/palantir/conjure/CachedPatternMatcher.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import java.time.Duration;
+import java.util.concurrent.ExecutionException;
+import java.util.regex.Pattern;
+
+public final class CachedPatternMatcher {
+    private final Pattern pattern;
+
+    @SuppressWarnings("checkstyle:IllegalType")
```
Comment (raiju) [inline conjure-generator-common/src/main/java/com/palantir/conjure/CachedPatternMatcher.java:29]: Discussion: do we prefer using guava caches or are we happy to add a dependency on Caffeine?

**Comment by carterkozak** [comment]
I tend to prefer unrolling regex into java code, that would avoid the cache and likely be just as fast.

**Comment by raiju** [comment]
To validate your suspicions, coded up the camel case matcher manually & ran some JMH benchmarks:
```
Benchmark                     Mode  Cnt   Score   Error   Units
RegexBenchmark.manual        thrpt    5  44.126 ± 0.614  ops/us
RegexBenchmark.matcherReuse  thrpt    5   1.286 ± 0.035  ops/us
RegexBenchmark.regex         thrpt    5   1.201 ± 0.005  ops/us
```

Looks like we can hit a > 40x improvement by coding it manually, which should help a lot.

**Comment by raiju** [comment]
@carterkozak Changed it to hand-written implementations

**Comment by carterkozak** [comment]
> Looks like we can hit a > 40x improvement by coding it manually, which should help a lot.

Looks great, thanks for working on this and validating with benchmarks :-)

File: conjure-generator-common/src/test/java/com/palantir/conjure/PatternTestUtil.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+package com.palantir.conjure;
+
+import java.util.List;
+import java.util.Random;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public final class PatternTestUtil {
```
Comment (carterkozak) [inline conjure-generator-common/src/test/java/com/palantir/conjure/PatternTestUtil.java:25]: It might be worthwhile to borrow from https://github.com/palantir/resource-identifier/blob/develop/src/test/java/com/palantir/ri/ResourceIdentifierPropertyTest.java

File: conjure-generator-common/src/main/java/com/palantir/conjure/CamelCasePattern.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+@Immutable
+public final class CamelCasePattern implements SimplifiedPattern {
+    @Override
+    public String pattern() {
+        return "^[a-z]([A-Z]{1,2}[a-z0-9]|[a-z0-9])*[A-Z]?$";
+    }
+
+    @Override
+    @SuppressWarnings("checkstyle:CyclomaticComplexity")
+    public boolean matches(String value) {
```
Comment (carterkozak) [inline conjure-generator-common/src/main/java/com/palantir/conjure/CamelCasePattern.java:30]: I'll need to set aside some time to review these carefully.

Comment (iamdanfox) [reply conjure-generator-common/src/main/java/com/palantir/conjure/CamelCasePattern.java:30]: @raiju when I hand-rolled a parser for sls-versions, I just put a 'jqwik' property-based testing to effectively fuzz the thing. https://github.com/palantir/sls-version-java/pull/455/files#diff-d90462e86f4e5e19c0df1b4545c7f9dcd7b440d389a391db7eca402d2f39b520R42

Helped catch the odd edge case for me 🤷

File: conjure-generator-common/src/test/java/com/palantir/conjure/PatternTestUtil.java (hunk 3 above)
Comment (schlosna) [reply conjure-generator-common/src/test/java/com/palantir/conjure/PatternTestUtil.java:25]: Inversely, might be worth writing hand rolled optimized RID parser/validator

File: conjure-generator-common/src/main/java/com/palantir/conjure/CamelCasePattern.java (hunk 4 above)
Comment (raiju) [reply conjure-generator-common/src/main/java/com/palantir/conjure/CamelCasePattern.java:30]: In the glorious habit of "not invented here", I added a little fuzzing test for each parser, comparing results w/ the regex: https://github.com/palantir/conjure/pull/999/files#diff-85b295230e0ec833cca21a2ab5b43c5015ec4436fbfacb263949786e4df0a45cR54

File: conjure-generator-common/src/test/java/com/palantir/conjure/PatternTestUtil.java (hunk 3 above)
Comment (carterkozak) [reply conjure-generator-common/src/test/java/com/palantir/conjure/PatternTestUtil.java:25]: That's exactly what that test is validating :-) It's ensuring that the old regex agrees with the new hand-rolled validator

Contributed here: https://github.com/palantir/resource-identifier/pull/140/

Comment (schlosna) [reply conjure-generator-common/src/test/java/com/palantir/conjure/PatternTestUtil.java:25]: Doh! I missed that PR landing for RIDs, it had popped up in some historic perf profiles I was looking at a while back. I know there are a few other places doing RID-like parsing that have popped in profiling that could use similar love.

Comment (carterkozak) [reply conjure-generator-common/src/test/java/com/palantir/conjure/PatternTestUtil.java:25]: Please let us know when those come up, happy to look into improving them :-)

File: conjure-generator-common/src/main/java/com/palantir/conjure/CamelCasePattern.java (hunk 4 above)
Comment (raiju) [reply conjure-generator-common/src/main/java/com/palantir/conjure/CamelCasePattern.java:30]: Moved it to the annotation-based one, and of course the limits of my little fuzzer came to light: my package parser was wrong for single-char packages, but my fuzzer only generated constant-length strings 🤦

File: conjure-generator-common/src/main/java/com/palantir/conjure/CamelCasePattern.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure;
+
+import com.google.errorprone.annotations.Immutable;
+
+@Immutable
+public final class CamelCasePattern implements SimplifiedPattern {
```
[the remark this answers is not in the document]
Comment (raiju) [reply conjure-generator-common/src/main/java/com/palantir/conjure/CamelCasePattern.java]: Done!

**Comment by raiju** [comment]
@carterkozak @iamdanfox Can I get a CR?

File: conjure-generator-common/src/main/java/com/palantir/conjure/CamelCasePattern.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+import com.google.errorprone.annotations.Immutable;
+
+@Immutable
+public enum CamelCasePattern implements SimplifiedPattern {
+    INSTANCE;
+
+    public static CamelCasePattern get() {
+        return INSTANCE;
+    }
```
Comment (carterkozak) [inline conjure-generator-common/src/main/java/com/palantir/conjure/CamelCasePattern.java:27]: why not use `INSTANCE` instead?

File: conjure-core/src/main/java/com/palantir/conjure/defs/validator/EnumValueDefinitionValidator.java [hunk 7]
```diff
@@ -56,14 +57,14 @@ public void validate(EnumValueDefinition definition) {
 
     @com.google.errorprone.annotations.Immutable
     private static final class FormatValidator implements ConjureValidator<EnumValueDefinition> {
-        private static final Pattern REQUIRED_FORMAT = Pattern.compile("[A-Z][A-Z0-9]*(_[A-Z0-9]+)*");
+        private static final SimplifiedPattern REQUIRED_FORMAT = EnumPattern.get();
```
Comment (carterkozak) [inline conjure-core/src/main/java/com/palantir/conjure/defs/validator/EnumValueDefinitionValidator.java:60]: imo this makes it easier to follow the code in an IDE:
```suggestion
        private static final EnumPattern REQUIRED_FORMAT = EnumPattern.get();
```

File: conjure-core/src/main/java/com/palantir/conjure/defs/validator/PackageValidator.java [hunk 8]
```diff
@@ -30,17 +31,15 @@
 
     private PackageValidator() {}
 
-    private static final Pattern VALID_PACKAGE = Pattern.compile("^([a-z][a-z0-9]+(\\.[a-z][a-z0-9]*)*)?$");
+    private static final SimplifiedPattern VALID_PACKAGE = PackagePattern.get();
```
Comment (carterkozak) [inline conjure-core/src/main/java/com/palantir/conjure/defs/validator/PackageValidator.java:34]: here as well:
```suggestion
    private static final PackagePattern VALID_PACKAGE = PackagePattern.INSTANCE;
```

File: conjure-generator-common/src/main/java/com/palantir/conjure/SimplifiedPattern.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+import com.google.errorprone.annotations.Immutable;
+
+/**
+ * A simplified/restricted pattern matcher. Goal is to execute faster than general regex at the cost of reduced
+ * features.
+ *
+ * JMH benchmarks suggest a ~40x increased throughput using the manual coded versions.
+ */
+@Immutable
```
Comment (carterkozak) [inline conjure-generator-common/src/main/java/com/palantir/conjure/SimplifiedPattern.java:27]: Interfaces cannot constrain mutability, nor can the iface itself be mutable
```suggestion
```

**Comment by carterkozak** [comment]
generally lgtm, just a couple of nits around redundant accessors

File: conjure-generator-common/src/main/java/com/palantir/conjure/SimplifiedPattern.java (hunk 9 above)
Comment (raiju) [reply conjure-generator-common/src/main/java/com/palantir/conjure/SimplifiedPattern.java:27]: We actually do need it due to the recursive checking of immutability (unless there's another trick to this)
```
/Volumes/git/conjure/conjure-generator-common/src/main/java/com/palantir/conjure/CaseConverter.java:74: warning: [ImmutableEnumChecker] enums should be immutable: 'Case' has field 'pattern' of type 'com.palantir.conjure.SimplifiedPattern', the declaration of type 'com.palantir.conjure.SimplifiedPattern' is not annotated with @com.google.errorprone.annotations.Immutable
        private final SimplifiedPattern pattern;
                                        ^
    (see https://errorprone.info/bugpattern/ImmutableEnumChecker)
error: warnings found and -Werror specified
1 error
1 warning
```

**Comment by raiju** [comment]
@carterkozak Addressed your comments

**Comment by carterkozak** [comment]
tyvm!
